File: ipycompss_lab_extension/src/start-stop/messaging.ts
import { Kernel, KernelMessage } from '@jupyterlab/services';
import { JSONObject } from '@lumino/coreutils';

import Utils from '../utils';

namespace StartStopMessaging {
  export interface IStartRequestDto extends JSONObject {
    arguments: JSONObject;
  }

  export interface ISuccessResponseDto extends JSONObject {
    success: boolean;
  }

  export const sendInitRequest = (
    kernel: Kernel.IKernelConnection | null | undefined
  ): Utils.IOnReply<ISuccessResponseDto> => {
    const initComm = kernel?.createComm('ipycompss_init_target');
    initComm?.open();
    return {
      onReply: (callback: (response: ISuccessResponseDto) => void) => {
        initComm &&
          (initComm.onMsg = (
            message: KernelMessage.ICommMsgMsg<'iopub' | 'shell'>
          ): void => callback(message.content.data as ISuccessResponseDto));
      }
    };
  };

  export const sendStartRequest = (
    kernel: Kernel.IKernelConnection | null | undefined,
    data: IStartRequestDto
  ): Utils.IOnReply<ISuccessResponseDto> => {
    const startComm = kernel?.createComm('ipycompss_start_target');
    startComm?.open(data);
    return {
      onReply: (callback: (response: ISuccessResponseDto) => void) => {
        startComm &&
          (startComm.onMsg = (
            message: KernelMessage.ICommMsgMsg<'iopub' | 'shell'>
          ): void => callback(message.content.data as ISuccessResponseDto));
      }
    };
  };

  export const sendStopRequest = (
    kernel: Kernel.IKernelConnection | null | undefined
  ): Utils.IOnReply<ISuccessResponseDto> => {
    const stopComm = kernel?.createComm('ipycompss_stop_target');
    stopComm?.open();
    return {
      onReply: (callback: (response: ISuccessResponseDto) => void) => {
        stopComm &&
          (stopComm.onMsg = (
            message: KernelMessage.ICommMsgMsg<'iopub' | 'shell'>
          ): void => callback(message.content.data as ISuccessResponseDto));
      }
    };
  };
}

export default StartStopMessaging;

File: ipycompss_kernel/ipycompss_kernel/kernel/start_stop/messaging.py
"""Methods for the start/stop messaging between the kernel and the front end"""
from typing import Any, Callable, TypedDict, Dict

import comm
from ipykernel.comm import Comm


class StartRequestDto(TypedDict):
    """Runtime start arguments"""

    arguments: Dict[str, Any]


class SuccessResponseDto(TypedDict):
    """Action success response"""

    success: bool


def on_init(callback: Callable[[], SuccessResponseDto]) -> None:
    """Register start message callback"""

    def on_init_comm(init_comm: Comm, _) -> None:
        response = callback()
        init_comm.send(data=response)
        del init_comm

    comm.get_comm_manager().register_target("ipycompss_init_target", on_init_comm)


def on_start(callback: Callable[[StartRequestDto], SuccessResponseDto]) -> None:
    """Register start message callback"""

    def on_start_comm(start_comm: Comm, open_start_comm: Dict[str, Any]) -> None:
        response = callback(open_start_comm["content"]["data"])
        start_comm.send(data=response)
        del start_comm

    comm.get_comm_manager().register_target("ipycompss_start_target", on_start_comm)


def on_stop(callback: Callable[[], SuccessResponseDto]) -> None:
    """Register stop message callback"""

    def on_stop_comm(stop_comm: Comm, _) -> None:
        response = callback()
        stop_comm.send(data=response)
        del stop_comm

    comm.get_comm_manager().register_target("ipycompss_stop_target", on_stop_comm)


def send_stop() -> None:
    """Send stop message to the frontend"""
    stop_comm: Comm = comm.create_comm("ipycompss_stop_target")
    del stop_comm
